opensquirrel/parsing/libqasm/libqasm_ir_creator.py

```diff
@@ -0,0 +1,130 @@
+import inspect
+import itertools
+
+import cqasm.v3x as cqasm
+
+from opensquirrel.default_gates import default_gate_aliases, default_gate_set
+from opensquirrel.gate_library import GateLibrary
+from opensquirrel.squirrel_ir import Float, Int, Qubit, SquirrelIR
+
+_cqasm_type_to_squirrel_type = {
+    cqasm.types.QubitArray: Qubit,
+    cqasm.values.ConstInt: Int,
+    cqasm.values.ConstReal: Float,
+}
+
+
+class LibqasmIRCreator(GateLibrary):
+    def __init__(self, gate_set=default_gate_set, gate_aliases=default_gate_aliases):
+        GateLibrary.__init__(self, gate_set, gate_aliases)
+        self.squirrel_ir = None
+
+    @staticmethod
+    def _get_qubits(cqasm_qubits_expression):
+        return [Qubit(index.value) for index in cqasm_qubits_expression.indices]
+
+    @staticmethod
+    def _get_literal(cqasm_literal_expression):
+        assert type(cqasm_literal_expression) in [cqasm.values.ConstInt, cqasm.values.ConstReal]
+
+        if type(cqasm_literal_expression) == cqasm.values.ConstInt:
+            return Int(cqasm_literal_expression.value)
+
+        if type(cqasm_literal_expression) == cqasm.values.ConstReal:
+            return Float(cqasm_literal_expression.value)
+
+    @staticmethod
+    def _check_cqasm_type(cqasm_expression, expected_squirrel_type):
+        cqasm_type = (
+            type(cqasm_expression.variable.typ)
+            if isinstance(cqasm_expression, cqasm.values.IndexRef)
+            else type(cqasm_expression)
+        )
+
+        if cqasm_type not in _cqasm_type_to_squirrel_type:
+            raise Exception("Unknown type")
+
+        if _cqasm_type_to_squirrel_type[cqasm_type] != expected_squirrel_type:
+            raise Exception("Unexpected type...")
+
+    @classmethod
+    def _get_expanded_squirrel_args(cls, generator_f, cqasm_args):
+        parameters = inspect.signature(generator_f).parameters.values()
+
+        ### Optional section: the below should never fire, as gate argument types are already checked by LibQasm.
+        if len(parameters) != len(cqasm_args):
+            raise Exception(f"Wrong number of arguments for gate {generator_f.__name__}")  # FIXME: aliases
```

Comment:
Since this is impossible to test (Libqasm raises an exception, so this code is never reached), I've removed the exception message (it wouldn't be user-facing) and replaced it with an assert